.github/workflows/main.yml

```diff
@@ -1,40 +1,72 @@
 ---
 name: CI
 
+env:
+  IMAGE_NAME: research-template
+  PUBLIC_IMAGE_NAME: ghcr.io/opensafely/research-template
+  REGISTRY: ghcr.io
+
 on:
+  workflow_dispatch:
+  pull_request:
   push:
+    branches: [main]
 
 jobs:
-  check:
+  build:
     runs-on: ubuntu-latest
-
     steps:
       - uses: actions/checkout@v4
+      - uses: hadolint/hadolint-action@54c9adbab1582c2ef04b2016b760714a4bfde3cf  # v3.1.0
+        # Lint the dockerfile before building
+        with:
+          failure-threshold: error
+          ignore: DL3022, DL3008, DL3009
       - uses: "opensafely-core/setup-action@v1"
         with:
-          python-version: "3.11"
           install-just: true
-      - name: Check formatting, linting and import sorting
-        run: just check
 
-  test:
-    runs-on: ubuntu-latest
+      - name: Build docker image
+        run: just build
+
+      - name: Save docker image
+        run: |
+          docker save research-template | gzip > /tmp/research-template.tar.gz
+
+      - name: Upload docker image
+        uses: actions/upload-artifact@v4
+        with:
+            name: research-template-image
+            path: /tmp/research-template.tar.gz
+
+  publish:
```

Comment:
question: If we're publishing on a successful build, *and* immediately making that the `latest` version, is it worth considering trying to add some very rudimentary smoke test ("does Python run? does R run?") before doing so?

Reply:
Of course, opensafely hasn't been installed yet. I've added a very basic test that we can extend later once we have the R installation working again.

Reply:
That's fine; even having something simple we can expand on is good :+1: